### ISSUE TITLE
How to configure multiple processes to use same GPU?

### ISSUE THREAD

**Issue by schuhschuh** [pytorch/ignite#2139]
## ❓ Questions/Help/Support

I have been using Ignite to distribute training over multiple GPUs on the same node. Here, each process is assigned a single dedicated GPU. I am wondering how I could have multiple processes use the same GPU? As in fractional GPU usage in Ray.Tune for example. As concrete example, what would I have to do in order to run 4 processes (`nproc_per_node=4`) with 2 GPUs, 2 processes per GPU? I think by default, the default CUDA device is set to `local_rank`. I suppose I somehow have to change the CUDA device, is there anything else to consider? Thanks for any tips.

**Comment by vfdev-5** [comment]
Interesting question, as far as I know we can not assign multiple processes to the same cuda device if we are using inter-proc communications with NCCL (can not find a reference for that):
```python
# CUDA_VISIBLE_DEVICES=0 python main.py
import time
import torch
import torch.distributed as dist
from torch.multiprocessing import start_processes

def main(rank):
    dist.init_process_group(
        "nccl", init_method="tcp://0.0.0.0:2233", world_size=2, rank=rank
    )
    torch.cuda.set_device(0)
    time.sleep(0.10 * rank)
    print(rank, ":", dist.get_world_size(), dist.get_backend())
    t = torch.tensor([0, 1, 2], device="cuda")
    dist.all_reduce(t)
    time.sleep(0.10 * rank)
    print(rank, ":", t)

    dist.destroy_process_group()

if __name__ == "__main__":
    start_processes(
        main, args=(), nprocs=2, start_method="spawn"
    )
```
>
```
RuntimeError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:912, invalid usage, NCCL version 20.9.8
ncclInvalidUsage: This usually reflects invalid usage of NCCL library (such as too many async ops, too many collectives at once, mixing streams in a group, etc).
```

However, this code works with "gloo" backend.

I can try a bit later one of our distributed examples to run like that: 2 procs on 1 GPU with gloo and can report here later.

HTH

**Comment by schuhschuh** [comment]
Interesting, thanks for looking into this so swift. I am now wondering if this even makes sense. I thought about it because I only used half the GPU memory. But having two processes with DistributedDataParallel on same GPU would probably be the same as doubling the batch size instead?! So I suppose the latter would be more reasonable in my current scenario.

For hyperparameter search with Ray.Tune it may still be something of interest. Ray supports it, so maybe this already works together with Ignite?

**Comment by vfdev-5** [comment]
> Interesting, thanks for looking into this so swift. I am now wondering if this even makes sense. I thought about it because I only used half the GPU memory. 

Computation part can be a bottleneck if using the same GPU. For example, if you launch the same training script on 1 GPU twice, training time for both processes will increase.

> But having two processes with DistributedDataParallel on same GPU would probably be the same as doubling the batch size instead?! 

Model will be copied for each process. As for the batch size you can still rescale it by the number of participating processes...

> So I suppose the latter would be more reasonable in my current scenario.

Maybe you can simply increase the batch size to take more GPU memory with a single process ? 

> For hyperparameter search with Ray.Tune it may still be something of interest. Ray supports it, so maybe this already works together with Ignite?

We had an feature request for thet here: https://github.com/pytorch/ignite/issues/1735
but haven't yet worked on that. I think it should be rather straighforward.

**Comment by schuhschuh** [comment]
Thanks! Indeed my original use case / intent was probably not meaningful.

As for Ray.Tune, I was just looking into how to use it with Ignite and came across the issue thread you linked. Looking forward to see how this can be done!

**Comment by ksasso1028** [comment]
I would like to acheive this to maximize the number of steps my model sees in a given time frame, this leading to faster convergence in my case. a larger batch size prolongs the number of steps I can accumulate in a given hour. @schuhschuh  @vfdev-5  @amyreese  @kostmo  @asford

**Comment by vfdev-5** [comment]
@ksasso1028 not sure to understand your question, sorry.
If you want to increase the number of steps, you can reduce the batch size, so you will have more model updates within a given timeframe.

**Comment by ksasso1028** [comment]
Reducing batch size increases steps yes, but greatly reduces GPU utilization. Meaning I could easily have additional processes running in parallel  on the dataset  to process parts of the set in parallel, leading to N times as many steps being taken while making full use of the GPU.  It is the same as using additional GPUs to process the dataset in parallel when the VRAM is full. Only difference is I want to do this on a single GPU, where each process does not take up the entire GPU.

**Comment by vfdev-5** [comment]
Right now I can think of the following hack using Gloo backend (nccl does not allow to address the same GPU from different processes), where we sort of running multi-node training using a single GPU:
```bash
#!/bin/bash

torchrun --nnodes=2 --node_rank=0 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=1 test_idist_docs_main.py &
torchrun --nnodes=2 --node_rank=1 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=1 test_idist_docs_main.py &
```

Python code:
```python
import time

import torch
from torchvision.models import resnet18

import ignite.distributed as idist


def training(local_rank, config):
    # ...
    time.sleep(0.5 * idist.get_rank())
    print(idist.get_rank(), ": run with config:", config, "- backend=", idist.backend(), idist.device())
    # ...
    model = resnet18(num_classes=10).to(idist.device())
    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

    model = idist.auto_model(model)
    
    x = torch.rand(512, 3, 32, 32, device=idist.device())
    y_pred = model(x)
    y_pred.mean().backward()

    time.sleep(0.5 * idist.get_rank())
    print(idist.get_rank(), ": ", next(model.parameters()).grad.mean())
    

backend = "gloo"
config = {}
with idist.Parallel(backend=backend) as parallel:
    parallel.run(training, config)

```

Output:
```
2024-05-15 23:26:56,315 ignite.distributed.launcher.Parallel INFO: Initialized processing group with backend: 'gloo'
2024-05-15 23:26:56,315 ignite.distributed.launcher.Parallel INFO: - Run '<function training at 0x7f4e1bc250d0>' in 2 processes
0 : run with config: {} - backend= gloo cuda:0
2024-05-15 23:26:56,599 ignite.distributed.auto.auto_model INFO: Apply torch DistributedDataParallel on model, device id: 0
1 : run with config: {} - backend= gloo cuda:0

0 :  tensor(8.3089e-05, device='cuda:0')
2024-05-15 23:26:57,713 ignite.distributed.launcher.Parallel INFO: End of run
2024-05-15 23:26:57,713 ignite.distributed.launcher.Parallel INFO: Finalized processing group with backend: 'gloo'
1 :  tensor(8.3089e-05, device='cuda:0')
```

This way you can imitate DDP with 2 processes using a single GPU.


Maybe, a proper solution could be to use cuda streams?

**Comment by ksasso1028** [comment]
great thank you! how would you do this sharding the dataloader per process instead of random input? I was hoping I could use something off the shelf but a custom implementation looks like it might be worth it in this case.

**Comment by ksasso1028** [comment]
nvm, just checked the docs. Haven't used ignite before, thanks for the help

**Comment by vfdev-5** [comment]
Dataloader sharding is done off the shelf using distributed sampler: https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler, if using ignite you can use `idist.auto_dataloader` to create DDP-aware dataloader: https://pytorch.org/ignite/generated/ignite.distributed.auto.auto_dataloader.html#ignite.distributed.auto.auto_dataloader.

You may already discovered this page: https://pytorch.org/ignite/distributed.html#distributed-launcher-and-auto-helpers

**Comment by ksasso1028** [comment]
thanks for the info. I have extended your example to 7 processes and its running great. There is a bit of work on my end to get my things working but was wondering, is it easy to scale this approach to N GPUs?

**Comment by vfdev-5** [comment]
> is it easy to scale this approach to N GPUs?

When you run the following with 2 fake nodes: 
```bash
torchrun --nnodes=2 --node_rank=0 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=1 test_idist_docs_main.py &
torchrun --nnodes=2 --node_rank=1 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=1 test_idist_docs_main.py &
```
there are 2 processes created with ranks: 0, 1 and local ranks: 0 (process of the first line), 0 (process of the second line) that's why second process is addressing the same GPU:0

If you now tweak `--nproc_per_node` parameter:
```bash
torchrun --nnodes=2 --node_rank=0 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=2 test_idist_docs_main.py &
torchrun --nnodes=2 --node_rank=1 --master_addr=0.0.0.0 --master_port=11223 --nproc_per_node=2 test_idist_docs_main.py &
```
then there are 4 processes created with ranks: 0, 1, 3, 4 and local ranks: 0, 1 (2 processes from the 1st line) and 0, 1 (2 processes from the 2nd line).
So, this way, 4 processes will be working on 2 GPUs.

**Comment by ksasso1028** [comment]
thanks, encountering a new issue. How can I pass the find_unused_parameters to ddp when running it with a module containing a submodule that is frozen (do not want to change)

RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).

**Comment by vfdev-5** [comment]
You can try to pass it into `auto_model` as kwargs:
```
model = idist.auto_model(model, find_unused_parameters=True)
```
https://pytorch.org/ignite/generated/ignite.distributed.auto.auto_model.html#ignite.distributed.auto.auto_model

**Comment by ksasso1028** [comment]
killing it @vfdev-5  thanks again. one more note, is there a way to kill the entire thing from the terminal other than using process ids and killing one by one? ctr + c had no affect on the bash script as soon as training started.

**Comment by vfdev-5** [comment]
Well, we have to google how to get pids automatically and kill all procs on Ctrl+c.
Thinking more about my previous messages, I have small doubt on DDP model wrapper whether it does expected thing. I'll check it later.

**Comment by ksasso1028** [comment]
look at that, learn something everyday. I have been testing it on some diffusion models and only thing I have noticed is the sampling procedure for outputs gets super slow when using 8 processes, I assume since it is being distributed across all processes.  Also realizing that GLOO is slower than NCCL, so I seem to be leaving speed on the table. With 8 processes, batch size of 16 (2 per process) I am getting ~24 seconds a epoch. Using typical lightning single process I get around 32 seconds per epoch using a batch size of 2. 
if you have any suggestions on speeding it up or a different route to take lmk.

**Comment by ksasso1028** [comment]
loss is also super bouncy in this setup, been tuning but wondering if I need to do any kind of a reduction when logging the loss on a end of epoch event?

`
    def train_step(engine, batch):
        # get clean sample
        net.train()
        dry = batch['dry']
        # dry_codes = self.net.encode(dry)
        # get features from network
        transformed, t, noise = transform(dry, rng)
        repair = net(transformed.to("cuda"), t.to("cuda"))
        l = loss(repair.to("cuda"),
                            noise.to("cuda"))
        l = l + 0. * sum(p.sum() for p in net.parameters())
        #l = engine.state.sync_all_reduce("batch_loss", l)
        l.backward()
        torch.nn.utils.clip_grad_norm_(net.parameters(), .2)
        optimizer.step()
        return l.detach()

trainer = Engine(train_step)
....more stuff

    tb_logger.attach_output_handler(
        trainer,
        event_name=Events.EPOCH_COMPLETED(every=1),
        tag="training",
        output_transform=lambda loss: {"batch_loss": loss},
    )
`

**Comment by vfdev-5** [comment]
>  Also realizing that GLOO is slower than NCCL, so I seem to be leaving speed on the table

well, this is correct. For real DDP config gloo is slower then nccl, but here we are using a fake DDP setup to address a single GPU from multiple processes. In this case nccl just can't work and gloo is working.

> if you have any suggestions on speeding it up or a different route to take lmk.

Use amp or do some profiling to check where the bottleneck: IO, data processing, model fwd/bwd passes ...
Otherwise, it would help to see the code in this case.

**Comment by ksasso1028** [comment]
Thanks, will profile and roundback. I am loading wav files from disk in the dataloader, and after scaling up to 8 processes the system gets slow almost unresponsive. Using atop I can see most of my memory being eaten up so thats a start. too bad bout nccl, nvidia wants you to purchase another GPU smh. 

btw heres the loss graphs, top is single process and bottom is the setup here in discussion with 8 processes. I am converging  faster as I can hear in my outputs as well. Seems like I can still tune the learning rate a bit more but glad to see its actually working faster with the same hardware. Each model being trained is running on a separate 3090 for comparision.
![image](https://github.com/pytorch/ignite/assets/11267645/6118c176-b41e-4d0b-b3f7-8e443d59284f)

**Comment by vfdev-5** [comment]
I think there can be a rule of thumb to apply to the learning rate when batch size grows : bs -> k * bs => lr -> k * lr

I was trying to check whether gradients in fake multi-node DDP 2 procs would correspond to real DDP 2 procs:
```python
import time

import torch
from torchvision.models import resnet18

import ignite.distributed as idist
import torch.distributed as dist
from ignite.utils import manual_seed


def pprint(rank, *msg):
    # We add sleep to avoid printing clutter
    time.sleep(0.5 * rank)
    print(rank, *msg)


def training(local_rank, config):
    # ...
    
    rank = idist.get_rank()
    device = idist.device()

    pprint(rank, ": run with config:", config, "- backend=", idist.backend(), idist.device())
    # ...

    
    # 1. Ensure all_reduce works:
    data_per_proc = torch.tensor([2.0 * rank + 0.5])
    # pprint(rank, f"data_per_proc = {data_per_proc}")
        
    data_per_proc = idist.all_reduce(data_per_proc)
    # pprint(rank, f"Now data_per_proc = {data_per_proc}")
    assert data_per_proc.item() == 3.0

    
    # 2. Ensure that grads are computed correctly:
    manual_seed(0)
    model = resnet18(num_classes=10).to(idist.device())
    optimizer = torch.optim.SGD(model.parameters(), lr=0.5)
    model = idist.auto_model(model)
    
    pprint(rank, ": before w", next(model.parameters()).mean().item())
    
    # manual_seed(2)
    # x = torch.rand(512, 3, 32, 32, device==idist.device())
    x = torch.arange(32 * 3 * 32 * 32, device=idist.device(), dtype=torch.float32).reshape(32, 3, 32, 32) / (32 * 3 * 32 * 32)

    optimizer.zero_grad()
    y_pred = model(x)
    y_pred.mean().backward()
    optimizer.step()

    pprint(rank, ": x:", x.mean().item())
    pprint(rank, ": after w", next(model.parameters()).mean().item())
    pprint(rank, ": w grad", next(model.parameters()).grad.mean().item())


backend = "gloo"
config = {}
with idist.Parallel(backend=backend) as parallel:
    parallel.run(training, config)
```
Results
    
```
Expected (DDP 2 procs on 2 GPUs):

0 : run with config: {} - backend= gloo cuda:0
1 : run with config: {} - backend= gloo cuda:1
2024-05-17 14:14:43,914 ignite.distributed.auto.auto_model INFO: Apply torch DistributedDataParallel on model, device id: 0
0 : before w -7.161527901189402e-05
1 : before w -7.161527901189402e-05


-- Random x:
0 : x: 0.5001485347747803
0 : after w 6.844229937996715e-05
0 : w grad -0.00028011511312797666

1 : x: 0.5005401372909546
1 : after w 6.844229937996715e-05
1 : w grad -0.00028011511312797666

-- Deterministic x:
0 : x: 0.4999949336051941
0 : after w 0.00671027647331357
0 : w grad -0.013563783839344978

1 : x: 0.4999949336051941
1 : after w 0.00671027647331357
1 : w grad -0.013563783839344978


Output (DDP 2 procs on a single GPU):

0 : run with config: {} - backend= gloo cuda:0
1 : run with config: {} - backend= gloo cuda:0
2024-05-17 14:15:23,804 ignite.distributed.auto.auto_model INFO: Apply torch DistributedDataParallel on model, device id: 0
0 : before w -7.161527901189402e-05
1 : before w -7.161527901189402e-05


-- Random x:
0 : x: 0.5001485347747803
0 : after w -0.0002313341392436996
0 : w grad 0.00031943770591169596

1 : x: 0.5001485347747803
1 : after w -0.0002313341392436996
1 : w grad 0.00031943770591169596

-- Deterministic x:
0 : x: 0.4999949336051941
0 : after w 0.0076806675642728806
0 : w grad -0.01550456415861845

1 : x: 0.4999949336051941
1 : after w 0.0076806675642728806
1 : w grad -0.01550456415861845
"""
```

**Comment by ksasso1028** [comment]
ouch, yea something with the optimization process certainly seems more bumpy in this setup. Do you have any other suggestions to get something to get parity with DDP on multi device? 

I have been contemplating an approach where I use python multi processing to kick off copies of the model with sharded data and run the epoch in parallel across the N processes. Instead of accumulating the gradients, I would backprop along each process and upon joining on the end of epoch, just average all the parameters together into a new network, rinse and repeat. Although unorthodox, removes the need to cuda processes to communicate. Whether a net would converge is another story.

**Comment by vfdev-5** [comment]
>  Do you have any other suggestions to get something to get parity with DDP on multi device?

maybe, it is not a big deal that they do not fully match in that example. There can be non-deterministic cuda behaviour involved...
